README.md
# node-bitstamp

bitstamp REST and WS API Node.js client :dollar:

## README Overview

* [Offers](#offers)
* [Example](#example)
* [Debug Info](#debug-info)
* [A word on parallel requests](#a-word-on-parallel-requests)
* [License](#license)

## Offers

* version 1 and version 2 of the bistamp REST API
* supports all new api endpoints
* 100% promise based
* optional full control of response status and headers
* all bitstamp currencies available: Bitcoin, Ethereum, Litecoin, Ripple, XRP
* also implements the websocket API v2 to listen for live trade, order book and live order events
* streams reconnect automatically and respect Forced reconnection via `bts:request_reconnect`
* takes care of signature and nonce automatically
* takes care of max. request quotas automatically (prevents bans)
* install via `npm i node-bitstamp` or `yarn add node-bitstamp`

## Example

* you can find a runnable api example [here](example/api.js), run via `yarn example:api`
* you can find a runnable stream example [here](example/stream.js), run via `yarn example:stream`

```javascript
"use strict";

const {BitstampStream, Bitstamp, CURRENCY} = require("node-bitstamp");

//printing available currencies
console.log(CURRENCY);

/* STREAMS */
// @ https://www.bitstamp.net/websocket/v2/

// Live trades
const bitstampStream = new BitstampStream();

/*
    as the stream is re-usable (subscribe to multiple currencies and channel types)
    every subscribe actions returns a channel name, which is the actual event you
    can listen to after subscription
*/

bitstampStream.on("connected", () => {
    const ethEurTickerChannel = bitstampStream.subscribe(bitstampStream.CHANNEL_LIVE_TRADES, CURRENCY.ETH_EUR);
    bitstampStream.on(ethEurTickerChannel, ({ data, event }) => {
        console.log(data);
        /* e.g.
            { 
                amount: 0.01513062,
                buy_order_id: 297260696,
                sell_order_id: 297260910,
                amount_str: '0.01513062',
                price_str: '212.80',
                timestamp: '1505558814',
                price: 212.8,
                type: 1,
                id: 21565524,
                cost: 3.219795936
            }
        */
    });
});
bitstampStream.on("disconnected", () => {});

// Live orderBook updates
bitstampStream.on("connected", () => {
    const btcEurOrderBookChannel = bitstampStream.subscribe(bitstampStream.CHANNEL_ORDER_BOOK, CURRENCY.BTC_EUR);

    bitstampStream.on(btcEurOrderBookChannel, ({ data, event }) => {
        console.log(data);
        /* e.g.
            { bids:
            [ 
                [ '3284.06000000', '0.16927410' ],
                [ '3284.05000000', '1.00000000' ],
                [ '3284.02000000', '0.72755647' ],
                .....
            ],
            asks:
            [ 
                [ '3289.00000000', '3.16123001' ],
                [ '3291.99000000', '0.22000000' ],
                [ '3292.00000000', '49.94312963' ],
                .....
            ] }
        */
    });
});

bitstampStream.on("error", (e) => console.error(e));

bitstampStream.close();

/* REST-API */
// @ https://www.bitstamp.net/api/

// @ https://www.bitstamp.net/account/login/
// To get an API key, go to "Account", "Security" and then "API Access". 
// Set permissions and click "Generate key"
// Dont forget to active the key and confirm the email.
const key = "abc3def4ghi5jkl6mno7";
const secret = "abcdefghijklmno";
const clientId = "123123";

const bitstamp = new Bitstamp({
    key,
    secret,
    clientId,
    timeout: 5000,
    rateLimit: true //turned on by default
});

const run = async () => {

    /*
        Every api function returns a bluebird promise.
        The promise only rejects on network errors or timeouts.
        A successfull promise always resolves in an object containing status, headers and body.
        status is the http status code as number, headers is an object of http response headers
        and body is the parsed JSON response body of the api, you dont need to parse the results
        yourself you can simply continue by accessing the object.
    */

    /* PUBLIC */
    const ticker = await bitstamp.ticker(CURRENCY.ETH_BTC).then(({status, headers, body}) => console.log(body));
    await bitstamp.tickerHour(CURRENCY.ETH_BTC);
    await bitstamp.orderBook(CURRENCY.ETH_BTC);
    await bitstamp.transactions(CURRENCY.ETH_BTC, "hour");
    await bitstamp.conversionRate();

    /* PRIVATE */
    const balance = await bitstamp.balance().then(({body:data}) => data);
    await bitstamp.userTransaction(CURRENCY.ETH_BTC, {offset, limit, sort});

    await bitstamp.openOrders(CURRENCY.ETH_BTC);
    await bitstamp.openOrdersAll();
    await bitstamp.orderStatus(id);
    await bitstamp.cancelOrder(id);
    await bitstamp.cancelOrdersAll();

    await bitstamp.buyLimitOrder(amount, price, currency, limit_price, daily_order);
    await bitstamp.sellLimitOrder(amount, price, currency, limit_price, daily_order);
    await bitstamp.buyMarketOrder(amount, currency);
    await bitstamp.sellMarketOrder(amount, currency);

    await bitstamp.withDrawalRequests(timedelta);
    await bitstamp.bitcoinWithdrawal(amount, address, instant);
    await bitstamp.bchWithdrawal(amount, address);
    await bitstamp.litecoinWithdrawal(amount, address);
    await bitstamp.ethereumWithdrawal(amount, address);
    await bitstamp.rippleWithdrawal(amount, address, currency);
    await bitstamp.xrpWithdrawal(amount, address, destination_tag);

    await bitstamp.bitcoinDepositAdress().then(({body}) => console.log(body));
    await bitstamp.bchDepositAdress().then(({body}) => console.log(body));
    await bitstamp.litecoinDepositAdress().then(({body}) => console.log(body));
    await bitstamp.ethereumDepositAdress().then(({body}) => console.log(body));
    await bitstamp.rippleDepositAdress().then(({body}) => console.log(body));
    await bitstamp.xrpDepositAdress().then(({body}) => console.log(body));

    await bitstamp.unconfirmedBitcoinDeposits();
    await bitstamp.transferSubToMain(amount, currency, subAccount);
    await bitstamp.transferMainToSub(amount, currency, subAccount);

    await bitstamp.openBankWithdrawal(/* {..} */);
    await bitstamp.bankWithdrawalStatus(id);
    await bitstamp.cancelBankWithdrawal(id);

    await bitstamp.newLiquidationAddress(currency);
    await bitstamp.liquidationAddressInfo(address);
};

run().then(() => {
    console.log(bitstamp.getStats());
    bitstamp.close();
});
```

## Debug Info

`DEBUG=node-bitstamp:* npm start`

## A word on parallel requests

* The client will never generate the same nonce for two requests.
* But a new request must always contain a higher nonce, than the last request before.
* When you make multiple calls in parallel (pretty easy in node..) there is a chance
    that the later calls reach the bitstamp api earlier than the first, causing the first
    requests to fail with an `invalid nonce` error.
* To prevent this you should make these calls sequentially.
* Besides chaining promises, this is another way to do it:

```javascript
const async = require("async"); // npm i async or yarn add async

async.series([
    cb => bitstamp.bitcoinDepositAdress()
        .then(r => cb(null, "BTC: " + r.body)).catch(e => cb(e)),
    cb => bitstamp.ethereumDepositAdress()
        .then(r => cb(null, "ETH: " + r.body.address)).catch(e => cb(e)),
    cb => bitstamp.litecoinDepositAdress()
        .then(r => cb(null, "LTC: " + r.body.address)).catch(e => cb(e))
], (error, data) => {

    if(error){
        return console.error(error);
    }

    console.log(data); // [ "BTC: ..", "ETH: ..", "LTC: .." ]
});
```

## License

MIT


CHANGELOG.md
# NODE-BITSTAMP CHANGELOG

# 3.1.0

- updated dependencies
- added CryptoTransactions
- added OHLC Data 

# 3.0.0

* support for new websockets
* new currencies supported
* bugfixes

# 2.3.0

* added btcusd

# 2.2.0

* updated dependencies
* typesript declarations
* fixed bug that encapsules api errors

## 2.1.0

* adding state of body by non 200 status codes into error message
* merged PR for liveorderstream

## 2.0.0

* small fixes
* **BREAKING** requests will now reject on bad status codes and api errors

## 1.7.0

* added BCH functions via PR
* fixed ep via PR

## 1.16.1

* fixed bug in bitcoinWithdrawal, instant is now an integer with default value 0

## 1.16.0

* added runnable examples
* updated docs
* repinned dependencies from ^ to ~
* upgraded dependencies to the latest version
* fixed small bugs in readme example code
* added cluster and encrypted configuration to the stream classes

## 1.15.0

* start of changelog

example/stream.js
"use strict";

const {BitstampStream, CURRENCY} = require("../index.js");

//printing available currencies
console.log(CURRENCY);

/* STREAMS */
// @ https://www.bitstamp.net/websocket/v2/

const bitstampStream = new BitstampStream();

bitstampStream.on("connected", () => {
    /*
        as the stream is re-usable (subscribe to multiple currencies / channels)
        every subscribe actions returns a channel name, which is the actual event you
        can listen to after subscription
    */
    const ethEurTickerChannel = bitstampStream.subscribe(bitstampStream.CHANNEL_LIVE_TRADES, CURRENCY.ETH_EUR);
    bitstampStream.on(ethEurTickerChannel, ({ data, event }) => {
        console.log(data);
        /* e.g.
            {
                amount: 0.01513062,
                buy_order_id: 297260696,
                sell_order_id: 297260910,
                amount_str: '0.01513062',
                price_str: '212.80',
                timestamp: '1505558814',
                price: 212.8,
                type: 1,
                id: 21565524,
                cost: 3.219795936
            }
        */
    });
});
bitstampStream.on("disconnected", () => {});

// Live orderBook updates
bitstampStream.on("connected", () => {
    const btcEurOrderBookChannel = bitstampStream.subscribe(bitstampStream.CHANNEL_ORDER_BOOK, CURRENCY.BTC_EUR);

    bitstampStream.on(btcEurOrderBookChannel, ({ data, event }) => {
        console.log(data);
        /* e.g.
            { bids:
            [
                [ '3284.06000000', '0.16927410' ],
                [ '3284.05000000', '1.00000000' ],
                [ '3284.02000000', '0.72755647' ],
                .....
            ],
            asks:
            [
                [ '3289.00000000', '3.16123001' ],
                [ '3291.99000000', '0.22000000' ],
                [ '3292.00000000', '49.94312963' ],
                .....
            ] }
        */
    });
});

bitstampStream.on("error", (e) => console.error(e));

setTimeout(() => {
    console.log("closing after 10 seconds.");
    bitstampStream.close();
}, 10000);
